### PR TITLE
Improve tabular datatypes.

### LINKED ISSUE #1818: File is incorrectly recognised as csv
The following file should not be recognised as `csv` but as `tabular` with one header.

https://github.com/galaxyproject/tools-iuc/blob/master/tools/datamash/test-data/group_compute_input.txt

### PR DESCRIPTION
**Pull request by jmchilton**
This PR merges the stale PR #1156 into the latest dev branch and introduces a hack into sniffing to fix #1818.

### CODE REVIEW

**Comment by nsoranzo** [comment]
@jmchilton The original bug description mentioned that "If backward compatibility is not required I will merge the new Excel_CSV datatype into the CSV datatype." What do you think about that?

**Comment by jmchilton** [comment]
@nsoranzo I'm indifferent on that point - I think the old datatype was probably new enough and broken enough that we shouldn't assume anyone sub-classed it - but I'm not sure I want to remove it as part of this PR either.

File: config/datatypes_conf.xml.sample [hunk 1]
```diff
@@ -52,9 +52,8 @@
       <converter file="interval_to_tabix_converter.xml" target_datatype="tabix" depends_on="bgzip"/>
       <converter file="bed_gff_or_vcf_to_bigwig_converter.xml" target_datatype="bigwig"/>
     </datatype>
-    <!-- MSI added Datatypes -->
-    <datatype extension="csv" type="galaxy.datatypes.tabular:CSV" display_in_upload="true" />
-    <!-- End MSI added Datatypes -->
+    <datatype extension="csv" type="galaxy.datatypes.tabular:ExcelCSV" display_in_upload="true" />
+    <datatype extension="tsv" type="galaxy.datatypes.tabular:ExcelCSV" display_in_upload="true" />
```
Comment (nsoranzo) [inline config/datatypes_conf.xml.sample]: s/ExcelCSV/ExcelTSV/

**Comment by nsoranzo** [comment]
@jmchilton I was thinking something like:
1. overwrite `CSV.sniff()` and `CSV.set_meta()` methods with the ones from `BaseCSV`
2. remove `BaseCSV` class and rename `CSV` to `BaseCSV`
3. rename `ExcelCSV` class to `CSV`

Does this make sense?

**Comment by martenson** [comment]
This might be the last blocking issue/PR for the 16.04 freeze - please consider pushing it back to 16.07.
@jmchilton

**Comment by jmchilton** [comment]
Sorry for holding up the release - I've changed the milestone - it fixes a bug so I can just open it against 16.04 anyway.

**Comment by nsoranzo** [comment]
:+1:

**Comment by martenson** [comment]
yay! thanks a lot @jmchilton & @nsoranzo & @Christian-B

**Comment by martenson** [comment]
this PR made it to freeze, changing its milestone

**Pull request merged**